File: Roadmap/01 - OPERADORES Y ESTRUCTURAS DE CONTROL/javascript/Roadmap/03 - ESTRUCTURAS DE DATOS/swift/didacdev.swift
import Foundation

//---------------- Array --------------------
/*
- Valores del mismo tipo ordenados.
- Puede incluir valores respetidos.
*/
let immutableArray: Array<String> // Immutable
var mutableArray: Array<String> = []
var arrayDos = Array(repeating: 0, count: 3)
var arrayTres = [0, 1, 2, 3]
var arrayCuatro = arrayTres + arrayDos

// Añadir
mutableArray.append("Hola")
print(mutableArray)
arrayTres += [4, 5, 6]
print(arrayTres)

// Obtener
print(arrayTres[0])

// Cambiar
mutableArray[0] = "adios"
print(mutableArray)
arrayTres[2...4] = [0, 0]
print(arrayTres)

// Insertar 
arrayTres.insert(4, at: 4)
print(arrayTres)

// Eliminar
arrayTres.remove(at: 2)
print(arrayTres)
arrayTres.removeLast()
print(arrayTres)
arrayTres.removeFirst()
print(arrayTres)
arrayTres.removeAll()
print(arrayTres)


//------------------- Set ------------
/*
- Distintos valores de un mismo tipo sin ordenar
*/
var emptySet = Set<String>()
var setDos: Set<Int> = [2, 1, 5, 4, 3]

// Añadir
emptySet.insert("Pepe")
print(emptySet)

// Eliminar
if let nombre = emptySet.remove("Pepe") {
    print(nombre)
} else {
    print("No existe")
}

// Ordenar
let ordenado = setDos.sorted()
print(ordenado)

//---------------- Dictionary -------------
/*
- Asociaciones clave-valor del mismo tipo
- Claves únicas
- No ordenados
*/

var emptyDict: [Int: String] = [:]
var dictDos = [2: "Pepe", 1: "Juan"]

// Actualizar
dictDos[1] = "Pedro"
print(dictDos)
dictDos.updateValue("Pablo", forKey: 2)
print(dictDos)

// Añadir
dictDos[0] = "Ana"
print(dictDos)

// Eliminar
dictDos.removeValue(forKey: 0)
print(dictDos)

print()

func agenda() {
    print("------------")
    print("    Agenda  ")
    print("------------")
    var agenda: [String: Int] = [:]
    let quit = false

    while(!quit) {
        print()
        print("1 - Buscar")
        print("2 - Insertar")
        print("3 - Actualizar")
        print("4 - Eliminar")

        if let entrada = readLine() {

            switch entrada {
                case "1":
                    search(agenda: &agenda)
                case "2":
                    insert(agenda: &agenda)
                case "3":
                    update(agenda: &agenda)
                case "4":
                    delete(agenda: &agenda)
                default:
                    print("Opción incorrecta.")
            }

        } else {
            print("No has introducido nada.")
        }
    }
}

func search(agenda: inout [String: Int]) {
    print()
    print("Nombre del contacto: ")

    if let contactName = readLine() {
        if let contact = agenda[contactName] {
            print("\(contactName): \(contact)")
        } else {
            print("No existe el contacto")
        }
    }
}

func insert(agenda: inout [String: Int]) {
    var name: String = ""
    var number: Int = 0

    print()
    print("Nombre del contacto: ")

    if let contactName = readLine() {
        name = contactName
    }
    
    print("Número del contacto: ")

    if let contactNumber = readLine() {
        if let numberInt = Int(contactNumber) {
            number = numberInt
        } else {
            print("Entrada incorrecta")
        }
    }

    agenda[name] = number

    print("\(name): \(number)")
}

func update(agenda: inout [String: Int]) {
    print()
    print("Nombre del contacto: ")

    if let contactName = readLine() {
        if let contactExist = agenda[contactName] {
            print("Nuevo número:")

            if let newNumber = readLine() {
                if let numberInt = Int(newNumber) {
                    agenda[contactName] = numberInt
                }
            } 
        } else {
            print("No existe el contacto")
        }
    }
}

func delete(agenda: inout [String: Int]) {
    print()
    print("Nombre del contacto")

    if let contactName = readLine() {
        if let contactExist = agenda[contactName] {
            agenda.removeValue(forKey: contactName)
        } else {
            print("El contacto no existe")
        }
    }
}

agenda()

File: Roadmap/03 - ESTRUCTURAS DE DATOS/kotlin/didacdev.kt
fun main() {

// ---------- Array ------------
/*
    - Número fijo de datos del mismo tipo
*/
    val array = arrayOf(0, 1, 2, 3, 4)

    // Obtener
    println(array[2])

// ----------- List -----------
/*
    - Almacena elementos en un orden especificado
*/
    val immutableList = listOf(0, 1, 2, 3, 4)
    val mutableList = mutableListOf(0, 3, 6, 2, 4)

    // Obtener
    println(immutableList.get(2))
    println(immutableList[2])

    // Añadir
    mutableList.add(1)

    // Eliminar
    mutableList.removeAt(2)

    // Actualizar
    mutableList[0] = 5

    println(mutableList)

    // Ordenar
    mutableList.sort()
    println(mutableList)

// -------- Set -------------
/*
    - Almacena elementos no repetidos
    - No tienen un orden definido
*/

    val immutableSet = setOf(1, 2, 3, 4)
    val mutableSet = mutableSetOf(2, 1, 5, 4)

    // Añadir
    mutableSet.add(3)

    // Eliminar
    mutableSet.remove(1)

    println(mutableSet)

    // Ordenar
    val ordenado = mutableSet.sorted()
    println(ordenado)

// ----------------- Map -------------
/*
    - Almacena pares clave-valor del mismo tipo
    - Claves únicas
*/
    val immutableMap = mapOf(1 to "pepe", 2 to "juan")
    val mutableMap = mutableMapOf(2 to "pepe", 1 to "juan")

    // Añadir
    mutableMap.put(3, "pedro")
    mutableMap[0] = "ana"

    // Actualizar
    mutableMap[1] = "paco"

    // Eliminar
    mutableMap.remove(0)

    println(mutableMap)

    // Ordenar
    val mapOrdenado = mutableMap.toSortedMap()

    println(mapOrdenado)

// -------------- ArrayDeque ----------------
/*
    - Implementación de cola y pila
    - permite añadir y eliminar el primer o último elemento
*/
    val deque = ArrayDeque(listOf(1, 2, 3, 4))

    // Añadir
    deque.addFirst(0)
    deque.addLast(5)

    println(deque)

    // Obtener
    println(deque.first())
    println(deque.last())

    // Eliminar
    deque.removeFirst()
    deque.removeLast()

    println(deque) 

    agenda()

}

fun agenda() {
    println("--------------")
    println("    Agenda    ")
    println("--------------")
    val agenda = mutableMapOf<String, Int>()
    var quit = false

    while(!quit) {
        println("")
        println("1 - Buscar")
        println("2 - Insertar")
        println("3 - Actualizar")
        println("4 - Eliminar")

        val entrada = readLine()

        if (entrada != null) {
            when (entrada) {
                "1" -> search(agenda)
                "2" -> insert(agenda)
                "3" -> update(agenda)
                "4" -> delete(agenda)
                else -> println("Opción incorrecta")
            }
        } else {
            println("No has introducido nada.")
        }
    }
}

fun search(agenda: MutableMap<String, Int>) {
    println("")
    println("Nombre del contacto:")

    val contactName = readLine()

    if (contactName != null) {
        val contact = agenda[contactName]
        
        if (contact != null) {

            println(" $contactName: $contact")
        } else {
            println("No existe el contacto")
        }
    } 
}

fun  insert(agenda: MutableMap<String, Int>) {
    var name: String = ""
    var number: Int

    println("")
    println("Nombre del contacto:")

    val contactName = readLine()

    if (contactName != null) {
        name = contactName
    } else {
        println("No existe el contacto")
    }

    println("Número del contacto: ")
    val contactNumber = readLine()
    val numberInt = contactNumber?.toInt()

    number = if (numberInt != null) numberInt else 0

    agenda[name] = number

    print("$name: $number")

}

fun update(agenda: MutableMap<String, Int>) {
    println("")
    println("Nombre del contacto:")

    val contactName = readLine()

    if (contactName != null) {
        if (agenda[contactName] != null){
            println("Nuevo número:")
            val newNumber = readLine()

            val numberInt = newNumber?.toInt()

            if (numberInt != null) {
                agenda[contactName] = numberInt
            }
        } else {
            println("No existe el contacto")
        }
    } else {
        println("No se ha introducido nada")
    }
}

fun delete(agenda: MutableMap<String, Int>) {
    println("")
    println("Nombre del contacto")

    val contactName = readLine()

    if (contactName != null) {
        if (agenda[contactName] != null) {
            agenda.remove(contactName)
        } else {
            println("No existe el contacto")
        }
    }
}
